IPs/RSKIP144.md

```diff
@@ -9,48 +9,88 @@
+![schedule](./RSKIP144/schedule.png)
 
-During execution, when a storage key is read, it is marked with the thread index in a per-partition readMap. When a key is written, the key is marked in a per-partition writeMap map. When all transactions of a certain partition have been processed, the writeMap is scanned. When all parallel partitions have finished processing, the readMaps and writeMaps are "merged". This requires efficient maps that enable traversing the keys in ascending lexicographic order. If a key belongs to a writeMap and a readMap of another partition, then the block is considered invalid. If a key belongs to a writeMap and a writeMap of another partition, then the block is also considered invalid. Recursive deletes must be correctly and efficiently handled. Miners are incentivized to produce a valid and efficient partition because by doing so their blocks spread faster over the network.
+## New block header field
 
-# Suggested miner implementation
+A new field `partitionEnds` is added to the block header. It determines how transactions are partitioned in a block.
+
+This field consists of an array of short unsigned integers that indicates at which position in the transaction list each partition ends. For example, in a block with 10 transactions, `partitionEnds = [3, 6]` indicates that the first _parallel partition_ contains transactions 0, 1 and 2; the second _parallel partition_ contains transactions 3, 4 and 5; and the _sequential partition_ contains transactions 6 to 9.
+
+The REMASC transaction must be included as the last transaction of the sequential partition.
+
+An empty `partitionEnds` indicates that all transactions go in the _sequential partition_.
+
+Values in `partitionsEnd` must be greater than 0 and in ascending order. An empty `partitionEnds` indicates that all transactions go in the sequential partition. The maximum number of parallel partitions that the miner can specify is equal to the minimum number of cores required to run the RSK node, this maximum is called  `maxTransactionExecutionThreads`.
+
+## New block validation consensus
+
+The new consensus mechanism allows to reject blocks that, when executed in parallel, can result in different outputs.
 
-When a miner executes transactions to create an (unsolved) block, the miner must execute the transactions serially and decide which partition the transaction belongs to. The block gas limit is replaced by a per-partition gas limit. We say that a transaction is *connected* to another transaction if:
+For simplicity, three types of connections between transactions are identified:
```

Comment:
Since "connections" are not something that actually exist between transactions, but something that we define to easily explain the algorithm, I would say that "We define a connection between two transactions if/when..."